sympy/core/operations.py

```diff
@@ -504,3 +505,118 @@ def args(self):
     @staticmethod
     def _compare_pretty(a, b):
         return (str(a) > str(b)) - (str(a) < str(b))
+
+class AssocOpDispatcher(OpDispatcher):
+    """
+    Handler dispatcher for associative operators
+
+    Explanation
+    ===========
+
+    If a new class is meant to return the result other than `Add` or `Mul` when operated,
+    it should have its handler defined as class method. Also, if the arguments have the same
+    ``_op_priority`` the priority relation of it and other classes should be registered in
+    this dispatcher by ``register_priority`` method.
+
+    Examples
+    ========
+
+    >>> from sympy import Add, Expr, Symbol
+    >>> from sympy.core.operations import AssocOpDispatcher
+    >>> newadd = AssocOpDispatcher("newadd", Add)
+
+    >>> class NewExpr(Expr):
+    ...     @classmethod
+    ...     def _newadd_handler(cls, *args, **kwargs):
+    ...         return NewAdd(*args, **kwargs)
+    >>> class NewAdd(NewExpr, Add):
+    ...     pass
+
+    >>> @newadd.register_priority(Expr, NewExpr)
+    ... @newadd.register_priority(NewExpr, Expr)
+    ... @newadd.register_priority(NewExpr, NewExpr)
+    ... def _(_1, _2):
+    ...     if issubclass(_1, NewExpr):
+    ...         return _1
+    ...     else:
+    ...         return _2
+
+    >>> a, b = Symbol('a'), NewExpr()
+    >>> newadd(a, a) == Add(a, a)
+    True
+    >>> newadd(a, b) == NewAdd(a, b)
+    True
+
+    Notes
+    =====
+
+    This approach is experimental, and can be replaced or deleted in the future.
+    See https://github.com/sympy/sympy/pull/19463.
+
+    """
+
+    def register(self, *args, **kwargs):
+        raise NotImplementedError("To avoid confusion, use 'register_priority' method instead.")
+
+    def register_priority(self, cls1, cls2):
+        """
+        Register the priority between two types. Registered function must
+        return the type which has the handler that is to be used.
+        """
+        def _(func):
+            self.add((cls1, cls2), func)
+            return func
+        return _
+
+    def binary_select(self, cls1, cls2):
+        """
+        Select the class using registered priority. If the priority is
+        not defined, return None.
+        """
+        result = self.dispatch(cls1, cls2)
+        if result is None:
+            return None
+        else:
+            return result(cls1, cls2)
+
+    def find_handler(self, args):
+        """
+        Select the argument with highest priority, and return its handler.
+        """
+        get = lambda arg: getattr(arg, self.handlermethod, None)
+        zargs = zip(args, map(get, args))
+        fzargs = filter(operator.itemgetter(1), zargs)
+
+        args_data = []
+        for a, handler in fzargs:
+            if handler is None:
+                continue
+            else:
+                func, op_priority = type(a), getattr(a, '_op_priority', 0)
+                args_data.append((func, op_priority, handler))
```

Comment:
The use of `map`/`zip`/`filter` here just seems to make things harder to read, and is probably slower. Why not:

```suggestion
        args_data = []
        for arg in args:
            handler = getattr(arg, self.handlermethod, None)
            if handler is None:
                continue
            func, op_priority = type(a), getattr(a, '_op_priority', 0)
            args_data.append((func, op_priority, handler))
```